vdj_ann/src/transcript.rs

```diff
@@ -7,187 +7,252 @@ use crate::refx::RefData;
-                    jstops.push((l + len) as i32);
-                }
-            }
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Default)]
+pub struct ContigStatus {
+    full_length: Option<bool>,
+    has_v_start: Option<bool>,
+    in_frame: Option<bool>,
+    no_premature_stop: Option<bool>,
+    has_cdr3: Option<bool>,
+    has_expected_size: Option<bool>,
+    correct_ann_order: Option<bool>,
+}
+
+impl ContigStatus {
+    fn is_productive(&self) -> bool {
+        match (
+            self.full_length,
+            self.has_v_start,
+            self.in_frame,
+            self.no_premature_stop,
+            self.has_cdr3,
+            self.has_expected_size,
+            self.correct_ann_order,
+        ) {
+            (
+                Some(true),
+                Some(true),
+                Some(true),
+                Some(true),
+                Some(true),
+                Some(true),
+                Some(true),
+            ) => true,
+            (_, _, _, _, _, _, _) => false,
         }
-        unique_sort(&mut vstarts);
-        unique_sort(&mut jstops);
-        let mut full = false;
-        for pass in 1..3 {
-            if pass == 2 && full {
-                continue;
-            }
-            let mut msg = "";
-            if pass == 2 {
-                msg = "frameshifted ";
-            };
-            for start in vstarts.iter() {
-                if !have_start(b, *start as usize) {
-                    continue;
-                }
-                for stop in jstops.iter() {
-                    let n = stop - start;
-                    if pass == 2 || n % 3 == 1 {
-                        let mut stop_codon = false;
-                        // shouldn't it be stop-3+1????????????????????????????????
-                        for j in (*start..stop - 3).step_by(3) {
-                            if have_stop(b, j as usize) {
-                                stop_codon = true;
-                            }
-                        }
-                        if !stop_codon {
-                            if pass == 1 {
-                                full = true;
-                            }
-                            if logme {
-                                fwriteln!(
-                                    log,
-                                    "{}full length transcript of length {}",
-                                    msg,
-                                    b.len()
-                                );
-                            }
-                        } else if logme {
-                            fwriteln!(
-                                log,
-                                "{}full length stopped transcript of length {}",
-                                msg,
-                                b.len()
-                            );
-                        }
-                    }
-                }
-            }
+    }
+
+    fn order_by(&self) -> u16 {
```

Comment:
Hmm, this method seems very odd.  What is your goal here?  If these criteria form some kind of ordered hierarchy, we might want to make that more explicit somehow.  Also, since it looks like most of these fields aren't actually optional, that will significantly change the logic of this method.

Reply:
A contig can be annotated with genes across multiple chains and we want to choose the status that is most informative. While these are arbitrary rules they match the logic of the original code and are going to need to be modified for different chemistries.